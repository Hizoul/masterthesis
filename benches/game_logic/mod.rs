pub mod tetromino;
pub mod field;
pub mod transposition;