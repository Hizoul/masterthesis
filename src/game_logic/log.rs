use serde::{Deserialize, Serialize};

#[derive(Debug,Serialize,Deserialize,Clone,Copy)]
pub struct Play {
    pub x: i8,
    pub rotation: i8,
    pub block: i8,
    pub index: u8,
    pub global_index: u8
}

#[derive(Debug,Serialize,Deserialize,Copy,Clone)]
pub enum Payload {
  PayloadRolled {
    from: i8,
    block: i8
  },
  PayloadPlaced {
    x: i8,
    y: i8,
    block: i8,
    orientation: i8,
    from: i8
  },
  PayloadSkipped {
    from: i8,
    block: i8,
    reason: i8
  },
  PayloadConsidering {
    play_index: u8
  }
}

#[derive(Debug,Serialize,Deserialize,Clone)]
pub struct GameLog {
    pub log: Vec<Payload>
}

impl GameLog {
  pub fn new() -> GameLog {
    return GameLog {
      log: Vec::with_capacity(100)
    }
  }
  pub fn add_item(&mut self, item: Payload) {
    self.log.push(item)
  }
  pub fn block_placed(&mut self, user: i8, block_type: i8, block_orientation: i8, x: i8, y: i8) {
    self.add_item(Payload::PayloadPlaced {
      from: user,
      block: block_type,
      orientation: block_orientation,
      x: x,
      y: y
    })
  }
  pub fn rolled_block(&mut self, user: i8, block_type: i8) {
    self.add_item(Payload::PayloadRolled {
      from: user,
      block: block_type
    })
  }
  pub fn player_skipped(&mut self, user: i8, block_type: i8, reason: i8) {
    self.add_item(Payload::PayloadSkipped {
      from: user,
      block: block_type,
      reason: reason
    })
  }
  pub fn add_consideration(&mut self, play_index: u8) {
    self.add_item(Payload::PayloadConsidering{
      play_index
    })
  }
  pub fn get_current_roll(&self) -> (i8, i8) {
    for e in self.log.iter().rev() {
      match e {
        Payload::PayloadRolled {block,from} => {
          return (*from, *block)
        }
        _ => {}
      }
    }
    return (0, 0)
  }
  pub fn copy_from(&mut self, other: &GameLog) {
    self.log.clear();
    for entry in other.log.iter() {
      match entry {
        Payload::PayloadConsidering {play_index} => {
          self.add_item(Payload::PayloadConsidering{
            play_index: *play_index
          })
        }
        Payload::PayloadPlaced {block,from,x,y,orientation} => {
          self.block_placed(*from, *block, *orientation, *x, *y);
        }
        Payload::PayloadSkipped {block,from, reason} => {
          self.player_skipped(*from, *block, *reason);
        }
        Payload::PayloadRolled {block,from} => {
          self.rolled_block(*from, *block);
        }
      }
    }
  }
}