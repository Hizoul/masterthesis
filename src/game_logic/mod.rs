pub mod val;
pub mod cache;
pub mod tetromino;
pub mod log;
pub mod field;