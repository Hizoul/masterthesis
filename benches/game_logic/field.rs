extern crate criterion;

use criterion::Criterion;
use rustyblocks::game_logic::field::{GameField};
use rustyblocks::game_logic::field::helper_structs::{GameRules};
use rustyblocks::game_logic::log::{GameLog};
use rustyblocks::game_logic::cache::{ShapeCache};
use std::sync::{Arc};
use rand::prelude::*;

const BLA: &'static str = "{\"log\":[{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":0,\"x\":8,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":0,\"x\":6,\"y\":0}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":0,\"x\":4,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":0,\"x\":7,\"y\":2}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":6,\"y\":2}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":7,\"y\":3}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":6,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":6,\"y\":10}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":6,\"y\":14}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":1,\"x\":7,\"y\":5}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":9,\"y\":3}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":3,\"x\":8,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":10,\"y\":6}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":2,\"y\":2}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":8,\"y\":7}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":2,\"x\":4,\"y\":2}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":7,\"y\":10}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":0,\"x\":4,\"y\":3}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":2,\"y\":6}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":3,\"x\":5,\"y\":5}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":10}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":3,\"x\":4,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":3,\"x\":8,\"y\":11}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":1,\"x\":4,\"y\":9}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":3,\"x\":7,\"y\":12}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":1,\"x\":0,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":4,\"y\":12}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":9,\"y\":12}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":14}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":3,\"x\":0,\"y\":3}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":0,\"y\":4}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":1,\"x\":4,\"y\":14}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":7,\"x\":0,\"y\":5}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":7,\"x\":1,\"y\":8}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":16}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":3,\"y\":4}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":3,\"x\":5,\"y\":15}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":3,\"y\":8}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":18}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":1,\"x\":2,\"y\":12}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":2,\"y\":15}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":7,\"y\":15}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":2,\"x\":3,\"y\":16}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":3,\"x\":0,\"y\":10}},{\"PayloadSkipped\":{\"from\":0,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":17}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":2,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":4,\"y\":17}},{\"PayloadSkipped\":{\"from\":0,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":2,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":2,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":19}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":2,\"x\":7,\"y\":18}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadConsidering\":{\"play_index\":0}}]}";


struct Foo {
  counter: usize
}

impl Foo {
  pub fn add(&mut self) {self.counter+=1;}
  pub fn get(&self) -> usize {self.counter}
}

trait FooTrait {
  fn t_add(&mut self);
  fn t_get(&self) -> usize;
}

impl FooTrait for Foo {
  fn t_add(&mut self) {self.add();}
  fn t_get(&self) -> usize {self.get()}
}

pub fn bench_field(c: &mut Criterion) {
  let cache = ShapeCache::new();
  let rules = GameRules::default();
  let p = Arc::new(cache);
  let p2 = p.clone();
  let p3 = p.clone();
  let log: GameLog = serde_json::from_str(BLA).unwrap();
  c.bench_function("deterministic_play_until_over", move |b| b.iter(|| {
    let mut f = GameField::new_with_cache(2, p.clone(), rules);
    while !f.game_over {
      f.place_block_using_play(0);
    }
  }));
  c.bench_function("random_play_until_over", move |b| b.iter(|| {
    let mut  f = GameField::new_with_cache(2, p2.clone(), rules);
    while !f.game_over {
      let p = f.possible_plays.len() as u8;
      f.place_block_using_play(rand::thread_rng().gen_range(0,p));
    }
  }));
  c.bench_function("replay_log", move |b| b.iter(|| {
    let mut f = GameField::new_with_cache(2, p3.clone(), rules);
    f.restore_from_log(&log, false);
  }));
  c.bench_function("traitcost_direct", move |b| b.iter(|| {
    let mut f = Foo {counter:0};
    for _ in 0..99999999 {
      while f.get() < 9999999999 {
        f.add();
      }
    }
  }));
  c.bench_function("traitcost_struct", move |b| b.iter(|| {
    let mut f: Box<dyn FooTrait> = Box::new(Foo {counter:0});
    for _ in 0..99999999 {
      while f.t_get() < 9999999999 {
        f.t_add();
      }
    }
  }));
}