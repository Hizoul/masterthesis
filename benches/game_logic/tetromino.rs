extern crate criterion;

use criterion::Criterion;
use rustyblocks::game_logic::tetromino::{Position,get_block_shape,get_translated_shape};
use rand::prelude::*;

pub fn bench_base_shape(c: &mut Criterion) {
  c.bench_function("block_shape", |b| b.iter(||
    get_block_shape(
      rand::thread_rng().gen_range(0,5),
      rand::thread_rng().gen_range(0,4)
    )));
}
pub fn bench_translated_shape(c: &mut Criterion) {
  c.bench_function("translate_shape", |b| b.iter(|| {
    let mut rng = rand::thread_rng();
    get_translated_shape(
      rng.gen_range(0,5),
      rng.gen_range(0,4),
      &Position::new(rng.gen_range(0,10),rng.gen_range(0,20))
    )
  }));
}