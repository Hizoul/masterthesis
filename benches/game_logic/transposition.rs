extern crate criterion;

use criterion::Criterion;
use rustyblocks::game_logic::field::{GameField};
use rustyblocks::game_logic::field::helper_structs::{GameRules};
use rustyblocks::game_logic::field::render::{Renderable};
use rustyblocks::zobrist::{ZobristHasher,U64Hasher};
use std::collections::HashMap;

pub fn bench_transpo(c: &mut Criterion) {
  c.bench_function("hash_u8", move |b| b.iter(|| {
    let mut set = HashMap::new();
    for _ in 0..3 {
      let mut f = GameField::new_with_rules(2, GameRules::deterministic());
      while !f.game_over {
        f.place_block_using_play(0);
        set.insert(f.to_field_string_with_player_as_u8(), 1);
      }
    }
  }));
  c.bench_function("hash_string", move |b| b.iter(|| {
    let mut set = HashMap::new();
    for _ in 0..3 {
      let mut f = GameField::new_with_rules(2, GameRules::deterministic());
      while !f.game_over {
        f.place_block_using_play(0);
        set.insert(f.to_field_string_with_player(true), 1);
      }
    }
  }));
  let mut zobrist_hasher2 = ZobristHasher::new(201, 20);
  c.bench_function("hash_zobrist_unopt", move |b| b.iter(|| {
    let hash_builder = U64Hasher::new();
    let mut set = HashMap::with_hasher(hash_builder);
    for _ in 0..3 {
      let mut f = GameField::new_with_rules(2, GameRules::deterministic());
      while !f.game_over {
        f.place_block_using_play(0);
        set.insert(zobrist_hasher2.hash(&f.to_field_string_with_player_as_u8()), 1);
      }
    }
  }));
  let mut zobrist_hasher3 = ZobristHasher::new(201, 20);
  c.bench_function("hash_zobrist_unopt_nou64hasher", move |b| b.iter(|| {
    let mut set = HashMap::new();
    for _ in 0..3 {
      let mut f = GameField::new_with_rules(2, GameRules::deterministic());
      while !f.game_over {
        f.place_block_using_play(0);
        set.insert(zobrist_hasher3.hash(&f.to_field_string_with_player_as_u8()), 1);
      }
    }
  }));
  let mut zobrist_hasher = ZobristHasher::new(201, 20);
  c.bench_function("hash_zobrist_field", move |b| b.iter(|| {
    let mut set = HashMap::new();
    for _ in 0..3 {
      let mut f = GameField::new_with_rules(2, GameRules::deterministic());
      while !f.game_over {
        f.place_block_using_play(0);
        set.insert(zobrist_hasher.hash_field(&f), 1);
      }
    }
  }));
}