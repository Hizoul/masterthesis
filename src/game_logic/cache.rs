use super::val::{GAMEFIELD_MAX_HEIGHT,GAMEFIELD_MAX_WIDTH,AMOUNT_OF_UNIQUE_PIECES_USIZE,AMOUNT_OF_UNIQUE_ORIENTATIONS_USIZE,GAMEFIELD_MAX_WIDTH_USIZE, GAMEFIELD_MAX_HEIGHT_USIZE,
AMOUNT_OF_UNIQUE_PIECES,AMOUNT_ROTATIONS};
use super::tetromino::{Tetromino,Position,BlockShape,get_translated_shape,get_block_shape};

#[derive(Debug,Clone)]
pub struct ShapeCache {
  base: Vec<Vec<BlockShape>>,
  translated: Vec<Vec<Vec<Vec<BlockShape>>>>
}

impl ShapeCache {
  pub fn new() -> ShapeCache {
    let mut base = Vec::with_capacity(AMOUNT_OF_UNIQUE_PIECES_USIZE);
    let mut translated = Vec::with_capacity(AMOUNT_OF_UNIQUE_PIECES_USIZE);
    for block in 0..AMOUNT_OF_UNIQUE_PIECES {
      let mut rotations = Vec::with_capacity(AMOUNT_OF_UNIQUE_ORIENTATIONS_USIZE);
      let mut translated_rotations = Vec::with_capacity(AMOUNT_OF_UNIQUE_ORIENTATIONS_USIZE);
      for rotation in 0..AMOUNT_ROTATIONS[block as usize] {
        rotations.push(get_block_shape(block, rotation));
        let mut xs = Vec::with_capacity(GAMEFIELD_MAX_WIDTH_USIZE+1);
        for x in 0..(GAMEFIELD_MAX_WIDTH) {
          let mut ys = Vec::with_capacity(GAMEFIELD_MAX_HEIGHT_USIZE+1);
          for y in 0..(GAMEFIELD_MAX_HEIGHT) {
            ys.push(get_translated_shape(block, rotation, &Position {x,y}));
          }
          xs.push(ys);
        }
        translated_rotations.push(xs)
      }
      base.push(rotations);
      translated.push(translated_rotations);
    }
    return ShapeCache {
      base, translated
    }
  }
  pub fn get_translated(&self, block: i8, rotation:i8, x: i8, y: i8) -> &BlockShape {
    self.translated.get(block as usize).expect("block available").get(rotation as usize).expect("rotation available").get(x as usize).expect("x available").get(y as usize).expect("y available")
  }
  pub fn get_base(&self, block: i8, rotation:i8) -> &BlockShape {
    self.base.get(block as usize).unwrap().get(rotation as usize).unwrap()
  }
  pub fn get_block_shape(&self, block: &Tetromino) -> &BlockShape {
    self.get_translated(block.block, block.rotation, block.position.x, block.position.y)
  }
}
