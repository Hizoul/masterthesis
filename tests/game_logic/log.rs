use rustyblocks::game_logic::log::{GameLog,Payload};
use insta::{assert_json_snapshot};

#[test]
fn gamelog_test() {
  let mut g: GameLog = GameLog::new();
  assert_json_snapshot!("before anything", &g);
  g.rolled_block(0, 2);
  assert_json_snapshot!("after roll", &g);
  g.block_placed(0, 1, 2, 3, 4);
  assert_json_snapshot!("after roll, place and roll", &g);
  g.rolled_block(1, 5);
  g.add_consideration(0);
  assert_json_snapshot!("created consideration", &g);
  for c in g.log.iter() {
    match &c {
      Payload::PayloadPlaced {from,x,y,block,orientation} => {
        assert_json_snapshot!("polaced", (from,x,y,block,orientation));
      }
      _ => {}
    }
  }
}