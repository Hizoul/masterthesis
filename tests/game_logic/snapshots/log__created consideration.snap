---
created: "2019-09-18T11:15:46.585308485Z"
creator: insta@0.8.1
source: tests/game_logic/log.rs
expression: "&g"
---
{
  "log": [
    {
      "PayloadRolled": {
        "from": 0,
        "block": 2
      }
    },
    {
      "PayloadPlaced": {
        "x": 3,
        "y": 4,
        "block": 1,
        "orientation": 2,
        "from": 0
      }
    },
    {
      "PayloadRolled": {
        "from": 1,
        "block": 5
      }
    },
    {
      "PayloadConsidering": {
        "play_index": 0
      }
    }
  ]
}
